libraries/persistent_store/src/concat.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Support for concatenated entries.
+//!
+//! This module permits to store multiple indexed values under the same key by concatenation. Such
+//! values must be at most 255 bytes and there can't be more than 255 such values under the same
+//! key (they are indexed with a `u8`).
+//!
+//! The rationale for using those particular constraints is that we want the number of bits to store
+//! the index and the number of bits to store the length to fit in an integer number of bytes
+//! (because the values are an integer number of bytes). Using only one byte is too restrictive
+//! (e.g. 8 values of at most 31 bytes or 16 values of at most 15 bytes). Using 2 bytes is plenty of
+//! space, so using one byte for each field makes parsing simpler and faster.
+//!
+//! The format is thus `(index:u8 length:u8 payload:[u8; length])*`. The concatenation is not
+//! particularly sorted.
+
+use crate::{Storage, Store, StoreError, StoreResult};
+use alloc::vec::Vec;
+use core::cmp::Ordering;
+use core::ops::Range;
+
+/// Reads a value from a concatenated entry.
+pub fn read(store: &Store<impl Storage>, key: usize, index: u8) -> StoreResult<Option<Vec<u8>>> {
+    let values = match store.find(key)? {
+        None => return Ok(None),
+        Some(x) => x,
+    };
+    Ok(find(&values, index)?.map(|range| values[range].to_vec()))
+}
+
+/// Writes a value to a concatenated entry.
+pub fn write(
+    store: &mut Store<impl Storage>,
+    key: usize,
+    index: u8,
+    value: &[u8],
+) -> StoreResult<()> {
+    if value.len() > 255 {
+        return Err(StoreError::InvalidArgument);
+    }
+    let mut values = store.find(key)?.unwrap_or(vec![]);
+    match find(&values, index)? {
+        None => {
+            values.push(index);
+            values.push(value.len() as u8);
+            values.extend_from_slice(value);
```

Comment:
Interesting, so it's actually a map. So if we know we will usually use increasing indices, we could also choose to not store the index. But this works for me, just wanted to point out that our use case probably doesn't need it.

Reply:
Yes, I thought about this but I actually think that we will eventually need it. Because we can't really ensure we have contiguous slots. What if a user deletes a slot in the middle? We would need to create a transaction with all slot-based entries to shift all slots by one. I think using a map is more robust to future changes.